### PR TITLE
Fix MPI issues in the well model

### PR DESCRIPTION
**Pull request by totto82**
The first commit makes some throws in `getWellResiduals` thread safe. 
The second make sure that the group related quantities are communicated before the wellTest

### CODE REVIEW

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
benchmark please

File: opm/simulators/wells/BlackoilWellModel_impl.hpp [hunk 1]
```diff
@@ -326,9 +326,9 @@ namespace Opm {
 
         well_state_ = previous_well_state_;
         well_state_.disableGliftOptimization();
+        updateAndCommunicateGroupData();
```
Comment (blattms) [inline opm/simulators/wells/BlackoilWellModel_impl.hpp:329]: Could the movement of this function to here, explain the failing tests?

**Review by blattms — COMMENTED** [review]
Apart from the failing parallel tests this seems good. Maybe we need to communicate group at the old place?

Once this tests are fixed this is fine to be merged.

File: opm/simulators/wells/BlackoilWellModel_impl.hpp (hunk 1 above)
Comment (totto82) [reply opm/simulators/wells/BlackoilWellModel_impl.hpp:329]: yes. but I think moving it here is correct. The old code used old group data in the welltest and well potential code etc.

**Comment by totto82** [comment]
> failing parallel tests

I don't see any of the parallel test failing. A update data should do here if you agree to the moving of `updateAndCommunicateGroupData()`

**Comment by bska** [comment]
> > failing parallel tests
> 
> I don't see any of the parallel test failing.

I agree with @totto82 .  The [failing tests](https://ci.opm-project.org/job/opm-simulators-PR-builder/1878/testReport/) are sequential regression tests only.  I suppose the `mpi.` in the test name (e.g., [mpi.compareECLFiles_flow+MOD4_GRP](https://ci.opm-project.org/job/opm-simulators-PR-builder/1878/testReport/(root)/mpi/compareECLFiles_flow_MOD4_GRP/)) "just" means that the tests are being run with an executable targeting an MPI configuration.  The simulation run however is effectively launched as
```sh
mpirun -np 1
```
Details of the process are coded in [compareECLFiles.cmake](https://github.com/OPM/opm-simulators/blob/831f1db872150fb218fa205f6827c14e261c6e6b/compareECLFiles.cmake) and [tests/run-regressionTest.sh](https://github.com/OPM/opm-simulators/blob/831f1db872150fb218fa205f6827c14e261c6e6b/tests/run-regressionTest.sh).

**Comment by blattms** [comment]
Indeed mpi indicates only compilation with mpi support. Sorry for the noise.

**Comment by totto82** [comment]
I have manually looked at the test failures. It should be safe to merge this with updated data.

**Comment by totto82** [comment]
jenkins build this update_data please

**Comment by bska** [comment]
jenkins build this opm-tests=491 please

**Comment by bska** [comment]
New reference solutions have been installed on the CI system.  I'll merge this now.  Thanks a lot for identifying this issue.

**Pull request merged**